npm/pkg/dataplane/dataplane_windows_test.go

```diff
@@ -0,0 +1,757 @@
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			hns := ipsets.GetHNSFake(t)
+			hns.Delay = defaultHNSLatency
+			io := common.NewMockIOShimWithFakeHNS(hns)
+			for ip, epID := range tt.ipEndpoints {
+				event := endpointCreateEvent(epID, ip)
+				require.Nil(t, event(nil, hns), "failed to create endpoint. ep ID: %s. ip: %s", epID, ip)
+			}
+
+			dp, err := NewDataPlane(thisNode, io, tt.cfg, nil)
+			require.NoError(t, err, "failed to initialize dp")
+
+			for k, event := range tt.events {
+				require.Nil(t, event(dp, hns), "failed while running event number %d", k)
+			}
+
+			verifyHNSCache(t, hns, tt.expectedSetPolicies, tt.expectedEnpdointACLs)
+
+			// uncomment to see output even for succeeding test cases
+			// require.Fail(t, "DEBUGME: successful")
+		})
+	}
+}
+
+func TestConcurrentEvents(t *testing.T) {
+	tests := []struct {
+		name                 string
+		cfg                  *Config
+		ipEndpoints          map[string]string
+		sessions             []concurrentSession
+		expectedSetPolicies  []*hcn.SetPolicySetting
+		expectedEnpdointACLs map[string][]*hnswrapper.FakeEndpointPolicy
+	}{
+		{
+			name:        "pod created on node -> concurrent: 1) creating relevant policy <=> 2) updating pod so policy no longer relevant",
+			cfg:         dpCfg,
+			ipEndpoints: nil,
+			sessions: []concurrentSession{
+				{
+					thread{
+						endpointCreateEvent(endpoint1, ip1),
+						podCreateEvent(applyDP, NewPodMetadata(podKey1, ip1, thisNode), ns1Set, podLabelSets1...),
+					},
+				},
+				{
+					thread{
+						policyUpdateEvent(policyNs1LabelPair1AllowAll()),
+					},
+					thread{
+						podUpdateEventSameIP(applyDP, NewPodMetadata(podKey1, ip1, thisNode), ns1Set, podLabelSets1, podLabelSets2),
+					},
+				},
+			},
+			expectedSetPolicies: []*hcn.SetPolicySetting{
+				setPolicy(emptySet),
+				setPolicy(allNamespaces, ns1Set.GetHashedName(), emptySet.GetHashedName()),
+				setPolicy(ns1Set, ip1),
+				setPolicy(podLabel2Set, ip1),
+				setPolicy(podLabelVal2Set, ip1),
+				// the rest are not garbage collected yet
+				setPolicy(podLabel1Set),
+				setPolicy(podLabel1Set),
+			},
+			expectedEnpdointACLs: map[string][]*hnswrapper.FakeEndpointPolicy{
+				endpoint1: {},
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			hns := ipsets.GetHNSFake(t)
+			hns.Delay = defaultHNSLatency
+			io := common.NewMockIOShimWithFakeHNS(hns)
+			for ip, epID := range tt.ipEndpoints {
+				event := endpointCreateEvent(epID, ip)
+				require.Nil(t, event(nil, hns), "failed to create endpoint. ep ID: %s. ip: %s", epID, ip)
+			}
+
+			dp, err := NewDataPlane(thisNode, io, tt.cfg, nil)
+			require.NoError(t, err, "failed to initialize dp")
+
+			for i, session := range tt.sessions {
+				// concurrent case with multiple go routines
+				wg := new(sync.WaitGroup)
+				wg.Add(len(session))
+				threadErrors := make(chan error, len(session))
+				for j, th := range session {
```

Comment:
since it was named as concurrent sessions, shouldn't each session start in separate process or thread?

Reply:
poor wording. thanks for pointing out